File: app/src/main/java/tpcreative/co/qrscanner/ui/scanner/ScannerPresenter.java
package tpcreative.co.qrscanner.ui.scanner;
import android.view.View;
import android.widget.LinearLayout;

import androidx.fragment.app.Fragment;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import tpcreative.co.qrscanner.R;
import tpcreative.co.qrscanner.common.SingletonHistory;
import tpcreative.co.qrscanner.common.Utils;
import tpcreative.co.qrscanner.common.controller.PrefsController;
import tpcreative.co.qrscanner.common.presenter.Presenter;
import tpcreative.co.qrscanner.common.services.QRScannerApplication;
import tpcreative.co.qrscanner.model.Create;
import tpcreative.co.qrscanner.model.EnumAction;
import tpcreative.co.qrscanner.model.EnumFragmentType;
import tpcreative.co.qrscanner.model.History;
import tpcreative.co.qrscanner.model.ItemNavigation;
import tpcreative.co.qrscanner.model.room.InstanceGenerator;
import tpcreative.co.qrscanner.ui.scannerresult.ScannerResultView;


public class ScannerPresenter extends Presenter<ScannerView>{

    protected List<Fragment> mFragment;
    protected HashMap<Object,String> hashClipboard = new HashMap<>();
    protected HashMap<Object,String>hashClipboardResult = new HashMap<>();
    protected StringBuilder stringBuilderClipboard = new StringBuilder();
    public ScannerPresenter(){
        mFragment = new ArrayList<>();
    }
    protected int mCount = 0;
    protected History history = new History();

    public void updateValue(int mCount){
        ScannerView view = view();
        this.mCount = this.mCount + mCount;
        view.updateValue(QRScannerApplication.getInstance().getString(R.string.total) +": "+ this.mCount);
    }


    public void doSaveItems(Create mCreate) {
       final Create create = mCreate;
        switch (create.createType){
            case ADDRESSBOOK:

                /*Put item to HashClipboard*/
                hashClipboard.put("fullName",create.fullName);
                hashClipboard.put("address",create.address);
                hashClipboard.put("phone",create.phone);
                hashClipboard.put("email",create.email);

                history = new History();
                history.fullName = create.fullName;
                history.address = create.address;
                history.phone = create.phone;
                history.email = create.email;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case EMAIL_ADDRESS:
                /*Put item to HashClipboard*/
                hashClipboard.put("email",create.email);
                hashClipboard.put("subject",create.subject);
                hashClipboard.put("message",create.message);

                history = new History();
                history.email = create.email;
                history.subject = create.subject;
                history.message = create.message;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case PRODUCT:
                /*Put item to HashClipboard*/
                hashClipboard.put("productId",create.productId);
                history = new History();
                history.text = create.productId;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case URI:
                /*Put item to HashClipboard*/
                hashClipboard.put("url",create.url);
                history = new History();
                history.url = create.url;
                history.createType = create.createType.name();
                onShowUI(create);
                break;

            case WIFI:
                /*Put item to HashClipboard*/
                hashClipboard.put("ssId",create.ssId);
                hashClipboard.put("password",create.password);
                hashClipboard.put("networkEncryption",create.networkEncryption);
                hashClipboard.put("hidden",create.hidden ? "Yes" : "No");
                history = new History();
                history.ssId = create.ssId;
                history.password = create.password;
                history.networkEncryption = create.networkEncryption;
                history.hidden = create.hidden;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case GEO:

                /*Put item to HashClipboard*/
                hashClipboard.put("lat",create.lat+"");
                hashClipboard.put("lon",create.lon+"");
                hashClipboard.put("query",create.query);
                history = new History();
                history.lat = create.lat;
                history.lon = create.lon;
                history.query = create.query;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case TEL:
                /*Put item to HashClipboard*/
                hashClipboard.put("phone",create.phone);
                history = new History();
                history.phone = create.phone;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case SMS:
                /*Put item to HashClipboard*/
                hashClipboard.put("phone",create.phone);
                hashClipboard.put("message",create.message);
                history = new History();
                history.phone = create.phone;
                history.message = create.message;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case CALENDAR:
                /*Put item to HashClipboard*/
                hashClipboard.put("title",create.title);
                hashClipboard.put("location",create.location);
                hashClipboard.put("description",create.description);
                hashClipboard.put("startEventMilliseconds", Utils.convertMillisecondsToDateTime(create.startEventMilliseconds));
                hashClipboard.put("endEventMilliseconds",Utils.convertMillisecondsToDateTime(create.endEventMilliseconds));
                history = new History();
                history.title = create.title;
                history.location = create.location;
                history.description = create.description;
                history.startEvent = create.startEvent;
                history.endEvent = create.endEvent;
                history.startEventMilliseconds = create.startEventMilliseconds;
                history.endEventMilliseconds = create.endEventMilliseconds;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            case ISBN:
                /*Put item to HashClipboard*/
                hashClipboard.put("ISBN",create.ISBN);
                history = new History();
                history.text = create.ISBN;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
            default:
                /*Put item to HashClipboard*/
                hashClipboard.put("text",create.text);
                history = new History();
                history.text = create.text;
                history.createType = create.createType.name();
                onShowUI(create);
                break;
        }
        try {
            final boolean autoCopy = PrefsController.getBoolean(QRScannerApplication.getInstance().getString(R.string.key_copy_to_clipboard),false);
            if (autoCopy){
                Utils.copyToClipboard(getResult(hashClipboard));
            }
        }
        catch (Exception e){
            e.printStackTrace();
        }
    }

    public void doRefreshView(){
        ScannerView view = view();
        mCount = 0;
        view.doRefreshView();
    }

    private void onShowUI(Create create){
        /*Adding new columns*/
        history.barcodeFormat = create.barcodeFormat;
        history.favorite = create.favorite;
        history.createDatetime = Utils.getCurrentDateTime();
        InstanceGenerator.getInstance(QRScannerApplication.getInstance()).onInsert(history);
        SingletonHistory.getInstance().reLoadData();
    }

    private String getResult(HashMap<Object,String> value){
        stringBuilderClipboard = new StringBuilder();
        if (value!=null && value.size()>0) {
            int i = 1;
            for (Map.Entry<Object, String> index : value.entrySet()) {
                if (i == value.size()) {
                    stringBuilderClipboard.append(index.getValue());
                } else {
                    stringBuilderClipboard.append(index.getValue());
                    stringBuilderClipboard.append("\n");
                }
                i += 1;
            }
            return stringBuilderClipboard.toString();
        }
        return "";
    }

    public void doShowAds(){
        if (Utils.isRelease()){
            return;
        }
        ScannerView view = view();
        if (Utils.isDebug() || Utils.isFreeRelease()){
            view.doShowAds(true);
        }else{
            view.doShowAds(false);
        }
    }
}

File: app/src/main/java/tpcreative/co/qrscanner/ui/scanner/ScannerPresenter.kt
package tpcreative.co.qrscanner.ui.scanner

import androidx.fragment.app.Fragment
import com.google.zxing.client.result.ParsedResultType
import tpcreative.co.qrscanner.R
import tpcreative.co.qrscanner.common.*
import tpcreative.co.qrscanner.common.controller.PrefsController
import tpcreative.co.qrscanner.common.presenter.Presenter
import tpcreative.co.qrscanner.common.services.QRScannerApplication
import tpcreative.co.qrscanner.helper.SQLiteHelper
import tpcreative.co.qrscanner.model.*
import java.util.*

class ScannerPresenter : Presenter<ScannerView?>() {
    protected var mFragment: MutableList<Fragment?>?
    protected var hashClipboard: HashMap<Any?, String?>? = HashMap()
    protected var hashClipboardResult: HashMap<Any?, String?>? = HashMap()
    protected var stringBuilderClipboard: StringBuilder? = StringBuilder()
    protected var mCount = 0
    protected var history: HistoryModel? = HistoryModel()
    fun updateValue(mCount: Int) {
        val view = view()
        this.mCount = this.mCount + mCount
        view.updateValue(QRScannerApplication.Companion.getInstance().getString(R.string.total) + ": " + this.mCount)
    }

    fun doSaveItems(mCreate: Create?) {
        when (mCreate.createType) {
            ParsedResultType.ADDRESSBOOK -> {
                /*Put item to HashClipboard*/hashClipboard["fullName"] = mCreate.fullName
                hashClipboard["address"] = mCreate.address
                hashClipboard["phone"] = mCreate.phone
                hashClipboard["email"] = mCreate.email
                history = HistoryModel()
                history.fullName = mCreate.fullName
                history.address = mCreate.address
                history.phone = mCreate.phone
                history.email = mCreate.email
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.EMAIL_ADDRESS -> {
                /*Put item to HashClipboard*/hashClipboard["email"] = mCreate.email
                hashClipboard["subject"] = mCreate.subject
                hashClipboard["message"] = mCreate.message
                history = HistoryModel()
                history.email = mCreate.email
                history.subject = mCreate.subject
                history.message = mCreate.message
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.PRODUCT -> {
                /*Put item to HashClipboard*/hashClipboard["productId"] = mCreate.productId
                history = HistoryModel()
                history.text = mCreate.productId
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.URI -> {
                /*Put item to HashClipboard*/hashClipboard["url"] = mCreate.url
                history = HistoryModel()
                history.url = mCreate.url
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.WIFI -> {
                /*Put item to HashClipboard*/hashClipboard["ssId"] = mCreate.ssId
                hashClipboard["password"] = mCreate.password
                hashClipboard["networkEncryption"] = mCreate.networkEncryption
                hashClipboard["hidden"] = if (mCreate.hidden) "Yes" else "No"
                history = HistoryModel()
                history.ssId = mCreate.ssId
                history.password = mCreate.password
                history.networkEncryption = mCreate.networkEncryption
                history.hidden = mCreate.hidden
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.GEO -> {

                /*Put item to HashClipboard*/hashClipboard["lat"] = mCreate.lat.toString() + ""
                hashClipboard["lon"] = mCreate.lon.toString() + ""
                hashClipboard["query"] = mCreate.query
                history = HistoryModel()
                history.lat = mCreate.lat
                history.lon = mCreate.lon
                history.query = mCreate.query
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.TEL -> {
                /*Put item to HashClipboard*/hashClipboard["phone"] = mCreate.phone
                history = HistoryModel()
                history.phone = mCreate.phone
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.SMS -> {
                /*Put item to HashClipboard*/hashClipboard["phone"] = mCreate.phone
                hashClipboard["message"] = mCreate.message
                history = HistoryModel()
                history.phone = mCreate.phone
                history.message = mCreate.message
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.CALENDAR -> {
                /*Put item to HashClipboard*/hashClipboard["title"] = mCreate.title
                hashClipboard["location"] = mCreate.location
                hashClipboard["description"] = mCreate.description
                hashClipboard["startEventMilliseconds"] = Utils.convertMillisecondsToDateTime(mCreate.startEventMilliseconds)
                hashClipboard["endEventMilliseconds"] = Utils.convertMillisecondsToDateTime(mCreate.endEventMilliseconds)
                history = HistoryModel()
                history.title = mCreate.title
                history.location = mCreate.location
                history.description = mCreate.description
                history.startEvent = mCreate.startEvent
                history.endEvent = mCreate.endEvent
                history.startEventMilliseconds = mCreate.startEventMilliseconds
                history.endEventMilliseconds = mCreate.endEventMilliseconds
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            ParsedResultType.ISBN -> {
                /*Put item to HashClipboard*/hashClipboard["ISBN"] = mCreate.ISBN
                history = HistoryModel()
                history.text = mCreate.ISBN
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
            else -> {
                /*Put item to HashClipboard*/hashClipboard["text"] = mCreate.text
                history = HistoryModel()
                history.text = mCreate.text
                history.createType = mCreate.createType.name
                onShowUI(mCreate)
            }
        }
        try {
            val autoCopy = PrefsController.getBoolean(QRScannerApplication.Companion.getInstance().getString(R.string.key_copy_to_clipboard), false)
            if (autoCopy) {
                Utils.copyToClipboard(getResult(hashClipboard))
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun doRefreshView() {
        val view = view()
        mCount = 0
        view.doRefreshView()
    }

    private fun onShowUI(create: Create?) {
        /*Adding new columns*/
        history.barcodeFormat = create.barcodeFormat
        history.favorite = create.favorite
        val time = Utils.getCurrentDateTimeSort()
        history.createDatetime = time
        history.updatedDateTime = time
        SQLiteHelper.onInsert(history)
        HistorySingleton.Companion.getInstance().reloadData()
    }

    private fun getResult(value: HashMap<Any?, String?>?): String? {
        stringBuilderClipboard = StringBuilder()
        if (value != null && value.size > 0) {
            var i = 1
            for ((_, value1) in value) {
                if (i == value.size) {
                    stringBuilderClipboard.append(value1)
                } else {
                    stringBuilderClipboard.append(value1)
                    stringBuilderClipboard.append("\n")
                }
                i += 1
            }
            return stringBuilderClipboard.toString()
        }
        return ""
    }

    init {
        mFragment = ArrayList()
    }
}